### DISCUSSION TITLE
mute myself over midi

### DISCUSSION THREAD

**Discussion by paulsijben** [jamulussoftware/jamulus#2322] [Ideas]
The current keyboard shortcut for mute myself works well but only when the Jamulus main window is the active one. Should you need to mute yourself when zoom, sibelius or any other bit of support software one is using, the keyboard shortcut is useless.
It would be great if a MIDI code can be used to achieve this as well.
 
I do not really mind which midi code needs to be used, but as people may be using the X-touch mini already to control incoming channels, CC code 87 or 88 as a default value might be useful.   (see also https://ianmorrish.files.wordpress.com/2019/08/xtouch-mini-mdi-mapping.png?w=1024)

**Comment by pljones** [comment]
This would need to fit into the current MIDI control structure (see the `--ctrlmidich` wiki help).

**Reply by paulsijben** [reply]
something like M87? No need for a count, right, otherwise ;1

**Reply by pljones** [reply]
I don't remember exactly how the parser works but yes, letter, cc num then, if needed, count.  But I'd hope, like you say, it could be omitted.

**Reply by henkdegroot** [reply]
So currently we have: 
f = volume fader; p = pan; m = mute; s = solo

I think to make this work we need to add an extra option in this list, as the mute myself is only a single item. All obvious letters for this are already in use. Perhaps the letter q (from quiet) could be an option?

I did not check the code yet to see how complex this is and if the current midi implementation can control the mute myself as well.
It sounds like a good option to have though.

**Reply by paulsijben** [reply]
@henkdegroot  or u for user ?

**Reply by henkdegroot** [reply]
It does not seem to be too much coding effort to get something like this working and from the looks of it, the "count" is not required to be provided.

**Reply by paulsijben** [reply]
I can not offer to do that coding, i can offer to do the testing!

**Reply by henkdegroot** [reply]
I will give it a try....hoping it is as easy as I believe it is.

**Reply by henkdegroot** [reply]
@paulsijben, you can download a test version here: https://github.com/henkdegroot/jamulus/actions/runs/1776820353

Currently the u character in the ctrlmidich option is used to set the CC# used for mute myself.

@pljones , do you know if midi control should work when using a headless client?

**Reply by paulsijben** [reply]
@henkdegroot the most issue I had was getting the midi patched in jackd . But now it works! Thank you. 
(BTW it would be great if the state would be sent back to the same MIDI CC with 127 for MUTED and 0 for unmuted but that is a nice to have ;-) )

However I am seeing that some midi events are lost and that coincides with errors like these
> ERROR: a2j_process_incoming: threw away MIDI event - not reserved at time 425927
> jack_midi_event_reserve: time parameter is earlier than last reserved event
I do not know if that is an issue with my Ubuntu setup or something in the Jamulus code???

**Reply by paulsijben** [reply]
oh and I double checked, it is not the MIDI controller sending a barrage of events

**Reply by henkdegroot** [reply]
> BTW it would be great if the state would be sent back

That would be very nice indeed, however the current midi implementation is one-way only.

Not sure about the "garbage" events, as Jamulus is just listening to the midi signals I don't think this can be caused by Jamulus,

**Reply by paulsijben** [reply]
the problem is that the events that are thrown away actually are good events that needed to be kept. So _too few_ events are being processed by Jamulus.

**Reply by pljones** [reply]
> @pljones , do you know if midi control should work when using a headless client?

Quite ridiculously, it's implemented in the Client Dialog class.  This needs fixing and is on my "list of things that should never have happened"... :)

It's all down to the current view of the architecture: the client can't be told to do anything except by the client dialog.  That's not right, really.

**Reply by henkdegroot** [reply]
> That's not right, really.

Any objections to extend to current implementation with 1 extra control in the dient dialog? 
And do you have an opinion about which character we should use in the ctrlmidich parameter? The u (from user) sounds logical.

**Reply by henkdegroot** [reply]
> ERROR: a2j_process_incoming: threw away MIDI event - not reserved at time 425927
> jack_midi_event_reserve: time parameter is earlier than last reserved event

Looks like this is "before" the midi data arrives in jack and can be picked up by Jamulus.

**Reply by pljones** [reply]
> Any objections to extend to current implementation with 1 extra control in the dient dialog?

Seems like the best option for now.

> And do you have an opinion about which character we should use in the ctrlmidich parameter? The u (from user) sounds logical.

```
  -M, --mutestream      starts the application in muted state
      --mutemyown       mute me in my personal mix (headless only)
```
It's called "mutemyown" from the command line (which is "headless only" - again, I have no idea _why_), so perhaps "o" for "own"?

I actually have _no_ idea what the "help" text for those two options _means_... so I could be wrong about what that option _does_...

**Reply by henkdegroot** [reply]
> I actually have _no_ idea what the "help" text for those two options _means_

I believe the mutestream option will enable "mute myself" at application startup, and the mutemyown will "mute" my channel in the mixer.

I like the "link" to own, so will do a PR which allows the letter o to be used to control the mute myself option. As long as we document this correct, then it should not be a problem.

**Reply by henkdegroot** [reply]
> ERROR: a2j_process_incoming: threw away MIDI event - not reserved at time 425927
> jack_midi_event_reserve: time parameter is earlier than last reserved event
> I do not know if that is an issue with my Ubuntu setup or something in the Jamulus code???

@paulsijben, I have done some testing using my virtual ubuntu setup and notice similar messages in a2jmidid. This seems to be some timing issue between a2j and JACK. I found that in JACK you can set the midi mode to RAW. With this option set you don't need to use a2jmidid as JACK will list the MIDI device without that as well. I did find (even in my virtual machine) that the response was better. The JACK messages dialog is showing XRUN error (so does not seem to handle all very well on my system). Perhaps using the RAW option makes it working better for you as well.

**Reply by paulsijben** [reply]
@henkdegroot thanks, will check later today!

**Reply by paulsijben** [reply]
how did you get the raw midi device to connect to the jamulus input? They seem to be different types in qjackctrl "purple" vs "red".

**Reply by henkdegroot** [reply]
> raw midi device to connect

Think the JACK version I used on Linux is a bit older version and I could just connect it. The version does not have the purple/red connections, just the 3 different tabs for Audio, Midi and Alsa. I launched another virtual machine and that one has a later version (with the Graph option) and indeed cannot connect the two midi ports.

**Reply by paulsijben** [reply]
check, I will re-test on my regular raspberry pi. Which code can I download/checkout to build a test version there?

**Reply by henkdegroot** [reply]
You can take the latest code from this PR: https://github.com/jamulussoftware/jamulus/pull/2334

**Reply by paulsijben** [reply]
got it to work on the PI. No errors from a2jmidid anymore!
However I did spot a little logic error in the new code! It currently mutes myself _while_ I press the button, but it does not toggle. So it now behaves like a cough rather than an on/off action. So if you ignore the CC message with value <0x40 (midi controller signalling button released), you will be OK

**Reply by pljones** [reply]
Or change the controller to send CC>=0x40 when pressed once, then CC<0x40 when pressed a second time, making the controller do the toggling?  (That's how my nanoKontrol1 can work - lights the LED to show the state.)

**Reply by paulsijben** [reply]
@pljones the example on the site deals with a Behringer X-touch mini, right? What does that one do? 
I know that a novation launchpad can not do it .

**Reply by henkdegroot** [reply]
@paulsijben, most devices allow you to define the buttons a "toggle" switches. Default they are "momentary" and this is causing the on+off behaviour you are experiencing. When you define the button as a "toggle" it will send ON the first time it is pressed and OFF the second time.

**Reply by paulsijben** [reply]
Just to check; how are we treating the mutes on the mixer?

**Reply by henkdegroot** [reply]
> treating the mutes on the mixer?

The exact same....

And according to a novation launchpad guide I found, you can: 

_The pads may behave either as toggles, triggers or momentary switches._

**Reply by paulsijben** [reply]
ah OK, then I shut up ;-)

**Reply by paulsijben** [reply]
thank you very much for building this!

**Reply by henkdegroot** [reply]
Thank you for asking this....it is a very nice thing to work on something like this. The question came across in another discussion as well, but as there was already a shortcut key for the function it stopped there. Now I got my own MIDI controller, I was keen on getting this in the software as well.

**Reply by paulsijben** [reply]
I made my own 1 button MIDI controller for this purpose especially, and updated the code to do the toggle in the button ;-)

Henk, should you be interested I can make you one too.

**Reply by henkdegroot** [reply]
> should you be interested I can make you one too.

I would be interested to know how you made it.....I have been looking at Palette Gear but found the price a bit high for my usage.

Another question would you see any need in being able to control the L + R PAN for your own input signal using MIDI?

**Reply by paulsijben** [reply]
I print this case https://www.thingiverse.com/thing:4810570 and then wrote a little bit of code to turn the pro micro into a midi controller. Added in some RGB LEDs to allow color feedback. 

I am not really using my own LR PAN usually I pan others in my mix so this is not something I see a use for on MIDI.

**Reply by henkdegroot** [reply]
Nice...going to look into that, it might be something nice to have

**Reply by paulsijben** [reply]
I checked my code into https://github.com/paulsijben/arduino if you want to build it yourself.

**Comment by softins** [comment]
Added to the upcoming 3.8.2 in #2334.
